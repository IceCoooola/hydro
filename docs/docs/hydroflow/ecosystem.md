---
sidebar_position: 4
---

# The Hydro Ecosystem
The Hydro Project is an evolving stack of libraries and languages for distributed programming.
A rough picture of the envisioned Hydro stack is below:

![Hydro Stack](./img/hydro_stack.png)

The core of the Hydro stack is shown in in the grey box; components that have not been implemented are in orange. 

Working up from the bottom:

- [Hydroplane](https://github.com/hydro-project/hydroplane) is a service for launching and monitoring Hydroflow transducers. It works both on local developer machines with Docker, and in cloud services like AWS EKS. Over time we expect to add autoscaling features to Hydroplane, to allow users to configure the cluster to grow and shrink based on the monitored behavior and a pluggable policy.

- [Hydroflow](https://github.com/hydro-project/hydroplane) is the subject of this book; a library for defining individual transducers in a distributed system. It uses the Rust compiler to generate binaries for deployment.

- *Hydrolysis* is a compiler we envision translating from Hydrologic to Hydroflow.

- *Hydrologic* is a high-level domain-specific language that we envision for distributed programming. Unlike Hydroflow, we expect Hydrologic to abstract away many notions of distributed computing. In particular, Hydrologic will be insensitive to the specific deployment of the codeâ€”the partitioning of functionality and data across transducers, the number of replicas of each transducer, etc. Instead, programmers will provide specifications for desired properties like the number of failures to tolerate, the consistency desired at a given endpoint, the latency of a given endpoint, etc. The Hydrolysis compiler will then generate Hydroflow transducers that can be deployed by Hydroplane to meet those specifications.

- [Metalift](https://github.com/metalift/metalift) is a framework for "lifting" code from one language to a (typically higher-level) language. We envision that Metalift will be used to translate from one of many distributed programming models/languages into our common Internal Representation, Hydrologic.

The Hydro stack is inspired by previous language stacks including [LLVM](https://llvm.org) and [Halide](https://halide-lang.org), which similarly expose multiple human-programmable Internal Representation langauges.

An early paper on the Hydro vision appeared in CIDR 2021, under the title [New Directions in Cloud Programming](https://www.cidrdb.org/cidr2021/papers/cidr2021_paper16.pdf).