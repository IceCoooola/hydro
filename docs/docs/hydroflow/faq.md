---
sidebar_position: 9
---

# FAQ
#### Q: How is Hydroflow different from dataflow systems like Spark, Flink, or MapReduce?
**A:** Hydroflow is designed as a substrate for building a wide range of distributed systems; traditional software dataflow systems
like the ones mentioned are targeted more narrowly at large-scale data processing. As such, Hydroflow differs from these systems in several ways:

First, Hydroflow is a lower-level abstraction than the systems mentioned. Hydroflow adopts a dataflow abstraction for specifying a transducer
running on a *single core*; one implementes a distributed system out of multiple Hydroflow transducers.  By contrast, Spark, Flink and MapReduce are 
distributed systems, which make specific choices for implementing issues like scheduling, fault tolerance, and so on.

Second, the systems mentioned above were designed specifically for distributed data processing tasks. 
By contrast, Hydroflow is designed as a compiler target for many possible language models. Distributed
dataflow systems like those mentioned above are one possible higher-level model one could build with Hydroflow, 
but there are many others.

Third, Hydroflow is a *reactive* dataflow engine, optimized for low-latency handling of 
asynchronous events and messages in a long-lived service. For example, one can build 
low-level protocol handlers in Hydroflow for implementing distributed protocols
like Paxos or Two-Phase Commit with low latency.

#### Q: What model of parallelism does Hydroflow use? SPMD? MPMD? Actors? MPI?
**A:** As a substrate for building individual nodes in a distributed systems, Hydroflow does not make any 
assumptions about the model of parallelism used. One can construct a distributed system out of Hydroflow 
transducers that use any model of parallelism, including 
[SPMD](https://en.wikipedia.org/wiki/Flynn%27s_taxonomy#Single_program,_multiple_data_streams_(SPMD)), 
[MPMD](https://en.wikipedia.org/wiki/Flynn%27s_taxonomy#Multiple_programs,_multiple_data_streams_(MPMD)), 
[Actors](https://en.wikipedia.org/wiki/Actor_model) or 
[MPI Collective Operations](https://en.wikipedia.org/wiki/Collective_operation). 
Hydroflow provides abstractions for implementing individual transducers to handle inputs, computing, and emitting outputs in
a general-purpose manner.

That said, it is common practice to deploy many instance of the same Hydroflow transducer; most distributed systems built in Hydroflow
therefore have some form of SPMD parallelism. (This is true of most distributed systems in general.)