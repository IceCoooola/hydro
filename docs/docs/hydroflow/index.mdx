---
sidebar_position: 1
---

import HydroflowDocs from '../../../hydroflow/README.md'

# Introduction

<HydroflowDocs />

## This Book
This book will teach you how to set up your environment to get started with Hydroflow, and how to program in the Hydroflow surface syntax.

Keep in mind that Hydroflow is under active development. However the code in this book is tested with the Hydroflow library so should always be up-to-date.

If you have any questions, feel free to [create an issue on Github](https://github.com/hydro-project/hydroflow/issues/new).
