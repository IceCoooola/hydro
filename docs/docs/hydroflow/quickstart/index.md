# Quickstart
This section will get you up and running with Rust and Hydroflow, and work you through a set of concrete examples illustrating much of what you need to know to get productive with Hydroflow.

:::tip

Hydroflow is the low-level IR that forms the foundation of the Hydro stack. If you're looking for a higher-level API, check out [Hydroflow+](../../hydroflow_plus/index.mdx).

:::
