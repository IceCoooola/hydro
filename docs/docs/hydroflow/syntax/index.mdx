import CodeBlock from '@theme/CodeBlock';
import exampleCode from '!!raw-loader!../../../../hydroflow/examples/example_syntax_hello_world.rs';

# Hydroflow Surface Syntax
The natural way to write a Hydroflow program is using the _Surface Syntax_ documented here.
It is a chained `Iterator`-style syntax of operators built into Hydroflow that should be sufficient
for most uses. If you want lower-level access you can work with the `Core API` documented in the [Architecture](../architecture/) section.

In this chapter we go over the syntax piece by piece: how to [embed surface syntax in Rust](./surface_embedding) and how to specify [_flows_](./surface_flows), which consist of [_data sources_](./surface_data) flowing through [_operators_](./surface_ops_gen).

<>{/* TODO(mingwei): In the [Hydroflow Types](surface_types.md) chapter we dive into the details of the data types that pass through flows. */}</>

As a teaser, here is a Rust/Hydroflow "HELLO WORLD" program:

<CodeBlock language="rust" showLineNumbers>{exampleCode}</CodeBlock>
