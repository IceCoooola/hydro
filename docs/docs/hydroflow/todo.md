---
sidebar_position: 8
---

# TODO

## Concepts
- p1 (Mingwei) Hydroflow and Rust: how do they go together?
    - State, control, scoping
- p1 State over time
    - lifetimes
    - explicit deletion
- p3 Coordination tricks?
    - End-of-stream to Distributed EOS?

## Docs
- p1 `hydroflow` struct and its methods
- p2 Review the ops docs

## Operators not discussed
    - dest_sink
    - identity
    - unzip
    - p1 *fold* -- add to chapter on [state](./syntax/state.md)
    - p1 *reduce* -- add to chapter on [state](./syntax/state.md)
    - p1 *fold_keyed* -- add to chapter on [state](./syntax/state.md)
    - p3 *sort_by_key* -- add to chapter on [state](./syntax/state.md)
    - p2 *next_stratum* -- add to chapter on [time](./concepts/life_and_times.md)
    - p2 *next_tick* -- add to chapter on [time](./concepts/life_and_times.md)
    - p2 *inspect* -- add to chapter on [debugging](./concepts/debugging.md)
    - p2 *null* -- add to chapter on [debugging](./concepts/debugging.md)

## How-Tos and Examples
- p1 Lamport clocks
- p2 Vector clocks
- p2 A partitioned Service
- p2 A replicated Service
- p2 Interfacing with external data
- p2 Interfacing with external services
- p1 Illustrate `'static` and `'tick` lifetimes (KVS)
- p3 Illustrate the `next_stratum` operator for atomicity (eg Bloom's upsert `<+-` operator)
- p3 Illustrate ordered streams (need `zip` operator ... what's the example?)
- p3 Actor model implementation (Borrow an Akka or Ray Actors example?)
- p3 Futures emulation? (Borrow a Ray example)
- p2 Illustrate external storage source and sink (e.g. for WAL of KVS)

## Odds and ends taken out of other chapters
- **Document the methods on the `hydroflow` struct** -- especially the run methods.
    -  The [`run_tick()`](https://hydro-project.github.io/hydroflow/doc/hydroflow/scheduled/graph/struct.Hydroflow.html#method.run_tick), [`run_stratum()`](https://hydro-project.github.io/hydroflow/doc/hydroflow/scheduled/graph/struct.Hydroflow.html#method.run_stratum), [`run()`](https://hydro-project.github.io/hydroflow/doc/hydroflow/scheduled/graph/struct.Hydroflow.html#method.run), and [`run_async()`](https://hydro-project.github.io/hydroflow/doc/hydroflow/scheduled/graph/struct.Hydroflow.html#method.run_async) methods provide other ways to control the graph execution.
    - Also `run_available()` `next_stratum()` and `recv_events` are important

- **Make sure `src/examples/echoserver` is the same as the template project** -- or better, find a way to do that via github actions or a github submodule

## What's covered in examples
- Concepts covered
    - cargo generate for templating
    - Hydroflow program specs embedded in Rust
    - Tokio Channels and how to use them in Hydroflow
        - Network sources and sinks (`source_stream`)
        - Built-in serde (`source_stream_serde`, `dest_sink_serde`)
    - Hydroflow syntax: operators, ->, variables, indexing multi-input/output operators
    - running Hydroflow via `run_available` and `run_async`
    - Recursion via cyclic dataflow
    - Fixpoints and Strata
    - Template structure: `clap`, message types
    - `source_stdin`
    - Messages and `demux`
    - broadcast pattern
    - the `persist` operator to store and replay dataflow
    - the `defer_signal` operator to gate a dataflow
    - bootstrapping pipelines: `initialize`

- Operators covered
    - cross_join
    - defer_signal
    - demux
    - dest_sink_serde
    - difference
    - filter
    - filter_map
    - flatten
    - flat_map
    - for_each
    - initialize
    - join
    - map
    - persist
    - union
    - source_iter
    - source_stdin
    - source_stream
    - source_stream_serde
    - tee
    - union
    - unique
