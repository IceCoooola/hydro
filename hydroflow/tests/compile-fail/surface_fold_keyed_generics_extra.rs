use hydroflow::hydroflow_syntax;

fn main() {
    let mut df = hydroflow::hydroflow_syntax! {
        source_iter(["hello", "world"])
            -> fold_keyed::<'tick, &str, String, ()>(String::new, |old: &mut _, val| {
                *old += val;
                *old += ", ";
            })
            -> for_each(|(k, v)| println!("({:?}, {:?})", k, v));
    };
    df.run_available();
}
