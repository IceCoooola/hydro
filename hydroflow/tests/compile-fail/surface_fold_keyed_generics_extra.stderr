error: `fold_keyed` should have at least 0 and at most 2 generic type arguments, actually has 3.
 --> tests/compile-fail/surface_fold_keyed_generics_extra.rs:6:29
  |
6 |             -> fold_keyed::<'tick, &str, String, ()>(String::new, |old: &mut _, val| {
  |                             ^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `hydroflow::hydroflow_syntax`
 --> tests/compile-fail/surface_fold_keyed_generics_extra.rs:1:5
  |
1 | use hydroflow::hydroflow_syntax;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
